onnxruntime/core/optimizer/transpose_optimizer/layout_tranformer_dev_notes.md

```diff
@@ -0,0 +1,55 @@
+# Layout Transformer Dev Notes
+
+## Overview
+ONNX standard assumes NCHW format for tensor layout. However, NCHW is not the best (perf efficient) format for all hardware types. Depending on the underlying hardware, to get the best perf we need to convert the model or in some cases part of the model from NCHW -> NHWC. Layout Transformer enables just this. It works with ONNX and ORT format models.
+
+*Note: Currently Layout Transformer works only for compiling EPs like NNAPI, QNN EP etc... More work is needed for it to be compatible with EPs like CPU and CUDA which use static kernel registration.*
+
+## How it works
+Layout transformer is [invoked during graph partitioning](https://github.com/microsoft/onnxruntime/blob/1a4868e5c4c4a270ad91036e36f2a03410c4c278/onnxruntime/core/framework/graph_partitioner.cc#L111) phase. After getting capabilities from the registered EP, layout transformer converts the part of the graph which is claimed by the EP to NHWC IF this is the desired format for the given EP.
+
+```
+Foreach
+  1. call GetCapability
+  2. IF EP.DesiredFormat == NHWC
+    2.1. Invoke Layout Transformer
+    2.2 If graph is modified -> call GetCapability (layout transformer can add new nodes to the graph)
+  3 Compile
+```
+
+GetCapability returns a bunch of IndexedSubGraphs that the given execution provider can run. During layout transformation, new nodes (Transpose, Gather etc) can be added within these subgraph. Therefore, calling GetCapability for the second time ensures that the EP can claim these new nodes as well and fuse the entire sub graphs. This is important for perf. Without this the execution will unnecessary switch to a fallback EP (in most cases CPU EP).
+
+*IMPORTANT NOTE* After layout transformation is done, graph resolve cannot be called for the graph. This is because graph resolve validates the shape of the nodes by calling ONNX TypeAndShapeInferenceFunction, these type and shape inf functions can ONLY infer shapes for NCHW format inputs. Therefore, when passed a graph with NHWC nodes the inferred shape validation fails and hence graph resolve throws. This is the very reason layout transformation is *NOT ENABLED* when Graph Partitioner Mode is kAssignOnly.
```

Comment:
```suggestion
*IMPORTANT NOTE* After layout transformation is done, Graph::Resolve cannot be called for the graph. This is because graph resolve validates the shape of the nodes by calling ONNX TypeAndShapeInferenceFunction, these type and shape inferencing functions can ONLY infer shapes for NCHW format inputs. Therefore, when passed a graph with NHWC nodes the inferred shape validation fails and hence graph resolve throws. This is the very reason layout transformation is *NOT ENABLED* when Graph Partitioner Mode is kAssignOnly.
```

Reply:
Double checking why Graph::Resolve fail. I would have expected the new nodes may not have an Op() given they're in the internal domain and an EP won't necessarily have a static kernel for the operator so you'd potentially hit a nullptr vs. the ONNX type/shape inferencing failing (which I'd expect to only apply to nodes in the ONNX domain).

Reply:
Updated this paragraph. Please re-review and let me know if it makes sense now.